packages/foam-vscode/src/features/create-from-template.ts

```diff
@@ -52,6 +52,105 @@ async function offerToCreateTemplate(): Promise<void> {
   }
 }
 
+function findFoamVariables(templateText: string): string[] {
+  const regex = /\$(FOAM_[_a-zA-Z0-9]*)|\${(FOAM_[[_a-zA-Z0-9]*)}/g;
+  var matches = [];
+  const output: string[] = [];
+  while ((matches = regex.exec(templateText))) {
+    output.push(matches[1] || matches[2]);
+  }
+  const uniqVariables = [...new Set(output)];
+  return uniqVariables;
+}
+
+function getFoamTitle(givenValues: Map<string, string>) {
+  if (givenValues.has('FOAM_TITLE')) {
+    return Promise.resolve(givenValues.get('FOAM_TITLE'));
+  } else {
+    return window.showInputBox({
+      prompt: `Enter a title for the new note`,
+      value: 'Title of my New Note',
+      validateInput: value =>
+        value.trim().length === 0 ? 'Please enter a title' : undefined,
+    });
+  }
+}
+
+function slugify(input: string) {
+  // TODO: Replace this with the actual algorithm.
+  return input.toLowerCase().replace(/[^\p{L}\p{N}]/gu, '-');
+}
+
+async function getFoamTitleSlug(
+  resolver: Resolver,
+  givenValues: Map<string, string>
+) {
+  if (givenValues.has('FOAM_TITLE_SLUG')) {
+    return Promise.resolve(givenValues.get('FOAM_TITLE_SLUG'));
+  }
+  const foamTitle = await resolver.resolve('FOAM_TITLE', givenValues);
+  return Promise.resolve(slugify(foamTitle));
+}
+
+class Resolver {
+  promises = new Map<string, Thenable<string>>();
+
+  resolve(name: string, givenValues: Map<string, string>): Thenable<string> {
+    if (!this.promises.has(name)) {
+      switch (name) {
+        case 'FOAM_TITLE': {
+          this.promises.set(name, getFoamTitle(givenValues));
+          break;
+        }
+        case 'FOAM_TITLE_SLUG': {
+          this.promises.set(name, getFoamTitleSlug(this, givenValues));
```

Comment:
have you considered explicitly making this `getFoamTitle` + slugification?

Reply:
Is the question whether to have a `getFoamTitleSlug` or inline the method here?

I like the symmetry/encapusulation of each foam variable having its implementation in its own method. 🤷 

---

(Aside: Part of me doesn't like the aesthetic of this switch statement at all and wants to remove it, but it hasn't made sense yet. Perhaps one day we'll be able to do something more dynamic/eval-y to make the code shorter/nicer: `get${variable}.call(this, givenValues)` or whatever the syntax is in JS.)

Reply:
I see as semantically different "getting" a variable from "derive a variable from another one", maybe that's why, although i agree there can be symmetry of implementation, I am not sure there is a symmetry of semantics.
E.g. imagine if when we create a new note from a wikilink we are automatically providing the `FOAM_TITLE` variable, but not the `FOAM_TITLE_SLUG` (we shouldn't and if we did it shouldn't be used); separating the two semantics would make this distinction very clear

In the end this is partially styling, I don't feel strongly about it, and I think the current approach is good, so I am happy to move forward with it.

Regarding your point about it being more dynamic, you could provide  have a map:

```
const resolvers = {
  FOAM_TITLE: getFoamTitle
  FOAM_TITLE_SLUG: getFoamTitleSlug
}
```

then check if the variable exists in the resolver table. Of course all functions would need to have the same signature but that sounds reasonable enough.
Connecting this to my previous point about symmetry, the map makes sense to me from an implementation POV, but to highlight the difference in semantics of the various methods I might name them differently, e.g. `getFoamTitle` and `computeFoamTitleSlug` 

Food for thoughts

Reply:
It might be pedantic, but:

* the code as it stands doesn't just "derive a variable from another one". It "resolves the value of `FOAM_TITLE`, **then** resolves a variable from it"
* IMO, the caller shouldn't care how the value is resolved, just that it is resolved. 

I've kept the naming generic, but changed `get` to `resolve` (ie. `resolveFoamTitle` and `resolveFoamTitleSlug`).